### PR TITLE
Update 'use collection expression' to support implicit reference conversions with arrays

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -47,8 +47,8 @@ private void AnalyzeArrayCreationExpression(SyntaxNodeAnalysisContext context, I
             return;
 
         // Analyze the statements that follow to see if they can initialize this array.
-        var allowInterfaceConversion = option.Value is CollectionExpressionPreference.WhenTypesLooselyMatch;
-        var matches = TryGetMatches(semanticModel, arrayCreationExpression, expressionType, allowInterfaceConversion, cancellationToken, out var changesSemantics);
+        var allowSemanticsChange = option.Value is CollectionExpressionPreference.WhenTypesLooselyMatch;
+        var matches = TryGetMatches(semanticModel, arrayCreationExpression, expressionType, allowSemanticsChange, cancellationToken, out var changesSemantics);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:51]: renamed `allowInterfaceConversion` to `allowSemanticsChange`.

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:51]: the bulk of this change is that rename.  i'll callout the actual code change.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs [hunk 2]
```diff
@@ -254,19 +254,39 @@ bool IsSafeConversionWhenTypesDoNotMatch(out bool changesSemantics)
             if (s_tupleNamesCanDifferComparer.Equals(type, convertedType))
                 return true;
 
-            if (allowInterfaceConversion &&
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs:257]: actual change is here.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs [hunk 3]
```diff
@@ -146,70 +146,6 @@ internal static class UseCollectionExpressionHelpers
 
         return true;
 
-        bool IsConstructibleCollectionType(ITypeSymbol? type)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs:149]: moved, unchanged to a public static helper.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs [hunk 4]
```diff
@@ -254,39 +190,127 @@ bool IsSafeConversionWhenTypesDoNotMatch(out bool changesSemantics)
...
+    }
+
+    public static bool IsWellKnownCollectionReadWriteInterface(ITypeSymbol type)
+    {
+        return type.OriginalDefinition.SpecialType
+            is SpecialType.System_Collections_Generic_ICollection_T
+            or SpecialType.System_Collections_Generic_IList_T;
+    }
+
+    public static bool IsConstructibleCollectionType(Compilation compilation, ITypeSymbol? type)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs:250]: just a move.  no other changes.

**Comment by CyrusNajmabadi** [comment]
@akhera99 this is ready for review.

**Review by akhera99 — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs [hunk 5]
```diff
@@ -204,6 +204,31 @@ class C
...
             LanguageVersion = LanguageVersion.CSharp12,
+            EditorConfig = """
+                [*]
+                dotnet_style_prefer_collection_expression=when_types_exactly_match
+                """
+        }.RunAsync();
+    }
+
+    [Fact]
+    public async Task TestNotWithIncompatibleExplicitArrays()
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForArrayTests.cs:215]: preference on a better name?

**Pull request merged**